tests/websocket.c

```diff
@@ -8,319 +8,537 @@
+    evbuffer_add(wt->wbuffer, "\x81", 1);
+    evbuffer_add(wt->wbuffer, &len, 1);
+    evbuffer_add(wt->wbuffer, mask, 4);
+    evbuffer_add(wt->wbuffer, cmd, sz);
+}
+
+static void
+ws_enqueue_frame(struct worker_thread *wt) {
+    char ping_command[] = "[\"PING\"]";
+    ws_enqueue_frame_for_command(wt, ping_command, sizeof(ping_command) - 1);
+
+    wait_for_possible_write(wt);
+}
+
+static int
+ws_on_message_complete(http_parser *p) {
+    struct worker_thread *wt = p->data;
+
+    wt->debug("%s (wt=%p)\n", __func__, wt);
+    // we've received the full HTTP response now, so we're ready to send frames
+    wt->state = WS_RECEIVED_HANDSHAKE;
+    ws_enqueue_frame(wt); /* add frame to buffer and register interest in writing */
+    return 0;
+}
+
+static void
+ws_on_timeout(evutil_socket_t fd, short event, void *arg) {
+    struct worker_thread *wt = arg;
+    (void) fd;
+    (void) event;
+
+    fprintf(stderr, "Time has run out! (thread %d)\n", wt->id);
+    event_base_loopbreak(wt->base); /* break out of event loop */
 }
 
 void*
 worker_main(void *ptr) {
 
-	char ws_template[] = "GET /.json HTTP/1.1\r\n"
-				"Host: %s:%d\r\n"
-				"Connection: Upgrade\r\n"
-				"Upgrade: WebSocket\r\n"
-				"Origin: http://%s:%d\r\n"
-				"Sec-WebSocket-Key1: 18x 6]8vM;54 *(5:  {   U1]8  z [  8\r\n"
-				"Sec-WebSocket-Key2: 1_ tx7X d  <  nw  334J702) 7]o}` 0\r\n"
-				"\r\n"
-				"Tm[K T2u";
-
-	struct worker_thread *wt = ptr;
-
-	int ret;
-	int fd;
-	struct sockaddr_in addr;
-	char *ws_handshake;
-	size_t ws_handshake_sz;
-
-	/* connect socket */
-	fd = socket(AF_INET, SOCK_STREAM, 0);
-	addr.sin_family = AF_INET;
-	addr.sin_port = htons(wt->hi->port);
-	memset(&(addr.sin_addr), 0, sizeof(addr.sin_addr));
-	addr.sin_addr.s_addr = inet_addr(wt->hi->host);
-
-	ret = connect(fd, (struct sockaddr*)&addr, sizeof(struct sockaddr));
-	if(ret != 0) {
-		fprintf(stderr, "connect: ret=%d: %s\n", ret, strerror(errno));
-		return NULL;
-	}
-
-	/* initialize worker thread */
-	wt->base = event_base_new();
-	wt->buffer = evbuffer_new();
-	wt->byte_count = 0;
-	wt->got_header = 0;
-
-	/* send handshake */
-	ws_handshake_sz = sizeof(ws_handshake)
-		+ 2*strlen(wt->hi->host) + 500;
-	ws_handshake = calloc(ws_handshake_sz, 1);
-	ws_handshake_sz = (size_t)sprintf(ws_handshake, ws_template, 
-			wt->hi->host, wt->hi->port,
-			wt->hi->host, wt->hi->port);
-	ret = write(fd, ws_handshake, ws_handshake_sz);
-
-	struct event ev_r;
-	event_set(&ev_r, fd, EV_READ | EV_PERSIST, websocket_read, wt);
-	event_base_set(wt->base, &ev_r);
-	event_add(&ev_r, NULL);
-
-	/* go! */
-	event_base_dispatch(wt->base);
-	event_base_free(wt->base);
-	free(ws_handshake);
-	return NULL;
+    char ws_template[] = "GET /.json HTTP/1.1\r\n"
+                "Host: %s:%d\r\n"
+                "Connection: Upgrade\r\n"
+                "Upgrade: WebSocket\r\n"
+                "Origin: http://%s:%d\r\n"
+                "Sec-WebSocket-Key: webdis-websocket-test-key\r\n"
```

Comment:
it would be good to validate the header sent back in response to this key.

Reply:
I've added this validation in commit 3984fb9.